File: apps/api/src/rx_jev_api/catalog.py
"""The v1 question catalog and which label sections may answer each question.

Code, not the model, decides the candidate sections. The mapper reads the sections a label
actually has, because product type metadata is not reliable: some repackager labels marked
prescription use the OTC Drug Facts layout, and older prescription labels predate the PLR
layout and use `warnings` and `precautions` instead of `warnings_and_cautions`.
"""

from typing import Literal

from pydantic import BaseModel

from rx_jev_api.clients.openfda import Label

__all__ = [
    "CATALOG",
    "Question",
    "QuestionId",
    "candidate_sections",
    "custom_sections",
    "get_question",
    "label_format",
]

QuestionId = Literal[
    "pregnancy",
    "breastfeeding",
    "children",
    "older_adults",
    "diabetes",
    "high_blood_pressure",
    "kidney",
    "liver",
    "heart",
    "asthma",
    "glaucoma",
    "enlarged_prostate",
    "stomach_ulcers",
    "alcohol",
    "blood_thinners",
    "drowsiness_driving",
    "take_with_food",
    "boxed_warning",
    "allergy",
]
Group = Literal["who", "conditions", "combinations", "daily_life", "serious"]
LabelFormat = Literal["otc", "prescription"]


class Question(BaseModel, frozen=True):
    id: QuestionId
    group: Group
    title: str
    # What Jev is asked about, phrased to fit "what does this label say about <subject>".
    # Question IDs are never sent to the model, so this carries the full meaning.
    subject: str
    # Candidate sections in priority order, per label layout. Old-layout prescription
    # fallbacks sit next to their PLR equivalents; absent sections are skipped.
    otc_sections: tuple[str, ...]
    prescription_sections: tuple[str, ...]
    # Extra rules for this question's stance, added to the shared reading rules.
    stance_rules: tuple[str, ...] = ()


# Sections that only appear in the OTC Drug Facts layout.
_OTC_MARKERS = {"do_not_use", "ask_doctor", "ask_doctor_or_pharmacist", "when_using", "stop_use"}
_OTC_MARKERS |= {"pregnancy_or_breast_feeding"}

_CONDITION_OTC = ("do_not_use", "ask_doctor", "warnings")
_CONDITION_RX = (
    "contraindications",
    "warnings_and_cautions",
    "warnings",
    "precautions",
    "use_in_specific_populations",
)


def _condition(id_: QuestionId, title: str, subject: str) -> Question:
    return Question(
        id=id_,
        group="conditions",
        title=title,
        subject=subject,
        otc_sections=_CONDITION_OTC,
        prescription_sections=_CONDITION_RX,
    )


CATALOG: tuple[Question, ...] = (
    Question(
        id="pregnancy",
        subject="use during pregnancy",
        group="who",
        title="Pregnancy",
        otc_sections=("pregnancy_or_breast_feeding",),
        prescription_sections=("pregnancy", "use_in_specific_populations"),
    ),
    Question(
        id="breastfeeding",
        subject="use while breastfeeding",
        group="who",
        title="Breastfeeding",
        otc_sections=("pregnancy_or_breast_feeding",),
        prescription_sections=("nursing_mothers", "use_in_specific_populations"),
    ),
    Question(
        id="children",
        subject="use in children",
        group="who",
        title="Children",
        otc_sections=("do_not_use", "dosage_and_administration"),
        prescription_sections=("pediatric_use",),
        # Agreed at Gate 1: unestablished safety tells parents something, so it is not silence.
        stance_rules=(
            "A statement that safety or effectiveness in children or pediatric patients has "
            "not been established is `caution`, not `not_mentioned`.",
        ),
    ),
    Question(
        id="older_adults",
        subject="use in older adults",
        group="who",
        title="Older adults",
        otc_sections=("ask_doctor", "dosage_and_administration"),
        prescription_sections=("geriatric_use",),
    ),
    _condition("diabetes", "Diabetes", "people who have diabetes"),
    _condition("high_blood_pressure", "High blood pressure", "people who have high blood pressure"),
    _condition(
        "kidney", "Kidney disease", "people who have kidney disease or reduced kidney function"
    ),
    _condition("liver", "Liver disease", "people who have liver disease"),
    _condition("heart", "Heart disease", "people who have heart disease, including heart failure"),
    _condition("asthma", "Asthma", "people who have asthma"),
    _condition("glaucoma", "Glaucoma", "people who have glaucoma"),
    _condition(
        "enlarged_prostate",
        "Enlarged prostate",
        "people who have an enlarged prostate or trouble urinating",
    ),
    _condition(
        "stomach_ulcers",
        "Stomach ulcers or bleeding",
        "people who have stomach ulcers or stomach bleeding",
    ),
    Question(
        id="alcohol",
        subject="drinking alcohol while using this drug",
        group="combinations",
        title="Alcohol",
        otc_sections=("warnings", "when_using"),
        prescription_sections=(
            "warnings_and_cautions",
            "warnings",
            "precautions",
            "drug_interactions",
        ),
    ),
    Question(
        id="blood_thinners",
        subject="using this drug together with blood thinners (anticoagulants such as warfarin)",
        group="combinations",
        title="Blood thinners",
        otc_sections=("ask_doctor_or_pharmacist",),
        prescription_sections=("drug_interactions",),
    ),
    Question(
        id="drowsiness_driving",
        subject="drowsiness, or driving and operating machinery, while using this drug",
        group="daily_life",
        title="Drowsiness and driving",
        otc_sections=("when_using",),
        prescription_sections=(
            "warnings_and_cautions",
            "warnings",
            "precautions",
            "information_for_patients",
        ),
    ),
    Question(
        id="take_with_food",
        subject="taking this drug with or without food",
        group="daily_life",
        title="Taking with food",
        otc_sections=("dosage_and_administration",),
        prescription_sections=("dosage_and_administration",),
    ),
    Question(
        id="boxed_warning",
        subject="serious risks highlighted in a boxed warning",
        group="serious",
        title="Boxed warning",
        otc_sections=(),
        prescription_sections=("boxed_warning",),
    ),
    Question(
        id="allergy",
        subject="people who have had an allergic reaction to this drug or its ingredients",
        group="serious",
        title="Allergy warnings",
        otc_sections=("do_not_use", "warnings"),
        prescription_sections=("contraindications",),
    ),
)

_BY_ID: dict[str, Question] = {q.id: q for q in CATALOG}


def get_question(question_id: str) -> Question:
    return _BY_ID[question_id]


def label_format(label: Label) -> LabelFormat:
    has_otc_marker = any(label.sections.get(marker, "").strip() for marker in _OTC_MARKERS)
    return "otc" if has_otc_marker else "prescription"


def candidate_sections(question_id: QuestionId, label: Label) -> list[str]:
    question = get_question(question_id)
    wanted = (
        question.otc_sections if label_format(label) == "otc" else question.prescription_sections
    )
    return [name for name in wanted if label.sections.get(name, "").strip()]


def custom_sections(label: Label) -> list[str]:
    """Sections a reader's own question may be answered from: every section the catalog
    reads for this label, in catalog order."""
    return list(dict.fromkeys(s for q in CATALOG for s in candidate_sections(q.id, label)))

File: apps/worker/src/catalog.ts
import type { Group, Layout } from '@rx-jev/contract'
import type { Label } from './clients/openfda.ts'
import { isBlank } from './python.ts'

// The v1 question catalog and which label sections may answer each question.
//
// Code, not the model, decides the candidate sections. The mapper reads the sections a label
// actually has, because product type metadata is not reliable: some repackager labels marked
// prescription use the OTC Drug Facts layout, and older prescription labels predate the PLR
// layout and use `warnings` and `precautions` instead of `warnings_and_cautions`.

export const QUESTION_IDS = [
  'pregnancy',
  'breastfeeding',
  'children',
  'older_adults',
  'diabetes',
  'high_blood_pressure',
  'kidney',
  'liver',
  'heart',
  'asthma',
  'glaucoma',
  'enlarged_prostate',
  'stomach_ulcers',
  'alcohol',
  'blood_thinners',
  'drowsiness_driving',
  'take_with_food',
  'boxed_warning',
  'allergy',
] as const
export type QuestionId = (typeof QUESTION_IDS)[number]

export type Question = {
  id: QuestionId
  group: Group
  title: string
  // What Jev is asked about, phrased to fit "what does this label say about <subject>".
  // Question IDs are never sent to the model, so this carries the full meaning.
  subject: string
  // Candidate sections in priority order, per label layout. Old-layout prescription
  // fallbacks sit next to their PLR equivalents; absent sections are skipped.
  otcSections: readonly string[]
  prescriptionSections: readonly string[]
  // Extra rules for this question's stance, added to the shared reading rules.
  stanceRules: readonly string[]
}

// Sections that only appear in the OTC Drug Facts layout.
const OTC_MARKERS = [
  'do_not_use',
  'ask_doctor',
  'ask_doctor_or_pharmacist',
  'when_using',
  'stop_use',
  'pregnancy_or_breast_feeding',
]

const CONDITION_OTC = ['do_not_use', 'ask_doctor', 'warnings']
const CONDITION_RX = [
  'contraindications',
  'warnings_and_cautions',
  'warnings',
  'precautions',
  'use_in_specific_populations',
]

function question(q: Omit<Question, 'stanceRules'> & { stanceRules?: string[] }): Question {
  return { stanceRules: [], ...q }
}

function condition(id: QuestionId, title: string, subject: string): Question {
  return question({
    id,
    group: 'conditions',
    title,
    subject,
    otcSections: CONDITION_OTC,
    prescriptionSections: CONDITION_RX,
  })
}

export const CATALOG: readonly Question[] = [
  question({
    id: 'pregnancy',
    subject: 'use during pregnancy',
    group: 'who',
    title: 'Pregnancy',
    otcSections: ['pregnancy_or_breast_feeding'],
    prescriptionSections: ['pregnancy', 'use_in_specific_populations'],
  }),
  question({
    id: 'breastfeeding',
    subject: 'use while breastfeeding',
    group: 'who',
    title: 'Breastfeeding',
    otcSections: ['pregnancy_or_breast_feeding'],
    prescriptionSections: ['nursing_mothers', 'use_in_specific_populations'],
  }),
  question({
    id: 'children',
    subject: 'use in children',
    group: 'who',
    title: 'Children',
    otcSections: ['do_not_use', 'dosage_and_administration'],
    prescriptionSections: ['pediatric_use'],
    // Agreed at Gate 1: unestablished safety tells parents something, so it is not silence.
    stanceRules: [
      'A statement that safety or effectiveness in children or pediatric patients has ' +
        'not been established is `caution`, not `not_mentioned`.',
    ],
  }),
  question({
    id: 'older_adults',
    subject: 'use in older adults',
    group: 'who',
    title: 'Older adults',
    otcSections: ['ask_doctor', 'dosage_and_administration'],
    prescriptionSections: ['geriatric_use'],
  }),
  condition('diabetes', 'Diabetes', 'people who have diabetes'),
  condition('high_blood_pressure', 'High blood pressure', 'people who have high blood pressure'),
  condition(
    'kidney',
    'Kidney disease',
    'people who have kidney disease or reduced kidney function',
  ),
  condition('liver', 'Liver disease', 'people who have liver disease'),
  condition('heart', 'Heart disease', 'people who have heart disease, including heart failure'),
  condition('asthma', 'Asthma', 'people who have asthma'),
  condition('glaucoma', 'Glaucoma', 'people who have glaucoma'),
  condition(
    'enlarged_prostate',
    'Enlarged prostate',
    'people who have an enlarged prostate or trouble urinating',
  ),
  condition(
    'stomach_ulcers',
    'Stomach ulcers or bleeding',
    'people who have stomach ulcers or stomach bleeding',
  ),
  question({
    id: 'alcohol',
    subject: 'drinking alcohol while using this drug',
    group: 'combinations',
    title: 'Alcohol',
    otcSections: ['warnings', 'when_using'],
    prescriptionSections: ['warnings_and_cautions', 'warnings', 'precautions', 'drug_interactions'],
  }),
  question({
    id: 'blood_thinners',
    subject: 'using this drug together with blood thinners (anticoagulants such as warfarin)',
    group: 'combinations',
    title: 'Blood thinners',
    otcSections: ['ask_doctor_or_pharmacist'],
    prescriptionSections: ['drug_interactions'],
  }),
  question({
    id: 'drowsiness_driving',
    subject: 'drowsiness, or driving and operating machinery, while using this drug',
    group: 'daily_life',
    title: 'Drowsiness and driving',
    otcSections: ['when_using'],
    prescriptionSections: [
      'warnings_and_cautions',
      'warnings',
      'precautions',
      'information_for_patients',
    ],
  }),
  question({
    id: 'take_with_food',
    subject: 'taking this drug with or without food',
    group: 'daily_life',
    title: 'Taking with food',
    otcSections: ['dosage_and_administration'],
    prescriptionSections: ['dosage_and_administration'],
  }),
  question({
    id: 'boxed_warning',
    subject: 'serious risks highlighted in a boxed warning',
    group: 'serious',
    title: 'Boxed warning',
    otcSections: [],
    prescriptionSections: ['boxed_warning'],
  }),
  question({
    id: 'allergy',
    subject: 'people who have had an allergic reaction to this drug or its ingredients',
    group: 'serious',
    title: 'Allergy warnings',
    otcSections: ['do_not_use', 'warnings'],
    prescriptionSections: ['contraindications'],
  }),
]

const BY_ID = new Map<string, Question>(CATALOG.map((q) => [q.id, q]))

export function getQuestion(questionId: string): Question {
  const found = BY_ID.get(questionId)
  if (found === undefined) throw new Error(`Unknown question: ${questionId}`)
  return found
}

function hasText(label: Label, section: string): boolean {
  return !isBlank(label.sections[section] ?? '')
}

export function labelFormat(label: Label): Layout {
  return OTC_MARKERS.some((marker) => hasText(label, marker)) ? 'otc' : 'prescription'
}

export function candidateSections(questionId: string, label: Label): string[] {
  const q = getQuestion(questionId)
  const wanted = labelFormat(label) === 'otc' ? q.otcSections : q.prescriptionSections
  return wanted.filter((name) => hasText(label, name))
}

/**
 * Sections a reader's own question may be answered from: every section the catalog reads
 * for this label, in catalog order.
 */
export function customSections(label: Label): string[] {
  return [...new Set(CATALOG.flatMap((q) => candidateSections(q.id, label)))]
}
